fennel/datasets/datasets.py

```diff
@@ -1700,6 +1707,184 @@ def sync_validation_for_pipelines(pipelines: List[Pipeline], ds_name: str):
                 tiers.add(tier)
 
 
+# ---------------------------------------------------------------------
+# Index
+# ---------------------------------------------------------------------
+
+
+class IndexType(str, Enum):
+    primary = "primary"
+
+    @classmethod
+    def _missing_(cls, value):
+        valid_types = [m.value for m in cls]
+        raise ValueError(
+            f"`{value}` is not a valid index 'type'. "
+            f"Following types are allowed {valid_types}"
+        )
+
+
+class IndexDuration(Enum):
+    none = None
+    forever = "forever"
+    duration = datetime.timedelta
+
+
+class Index(BaseModel):
+    type: IndexType
+    online: bool
+    offline: IndexDuration
+
+
+@overload
+def index(  # noqa: E704
+    *,
+    type: Literal["primary"] = DEFAULT_INDEX_TYPE,  # type: ignore
+    online: bool = DEFAULT_INDEX_ONLINE,
+    offline: Optional[str] = DEFAULT_INDEX_OFFLINE,
+) -> Callable[[Dataset], Dataset]: ...
+
+
+@overload
+def index(obj: Dataset) -> Dataset: ...  # noqa: E704
+
+
+def index(
+    obj: Dataset = None,
+    type: Literal["primary"] = DEFAULT_INDEX_TYPE,  # type: ignore
+    online: bool = DEFAULT_INDEX_ONLINE,
+    offline: Optional[str] = DEFAULT_INDEX_OFFLINE,
+) -> Union[Callable[[Dataset], Dataset], Dataset]:
+    """index decorator, defaults to a dataset with primary index, online set to true and forever lookup.
```

Comment:
and offline index is forever retention ?

Reply:
yup